src/transport/SessionManager.cpp

```diff
@@ -773,29 +840,54 @@ void SessionManager::SecureGroupMessageDispatch(const PacketHeader & packetHeade
         return;
     }
 
-    System::PacketBufferHandle msgCopy;
+    // Extract MIC from the end of the message.
+    uint8_t * data     = msg->Start();
+    uint16_t len       = msg->DataLength();
+    uint16_t footerLen = packetHeader.MICTagLength();
+    VerifyOrReturn(footerLen <= len);
+
+    uint16_t taglen = 0;
+    MessageAuthenticationCode mac;
+    ReturnOnFailure(mac.Decode(packetHeader, &data[len - footerLen], footerLen, &taglen));
+    VerifyOrReturn(taglen == footerLen);
+
     bool decrypted = false;
     while (!decrypted && iter->Next(groupContext))
     {
-        // Optimization to reduce number of decryption attempts
-        if (groupId != groupContext.group_id)
+        CryptoContext context(groupContext.keyContext);
+        msgCopy = msg.CloneData();
```

Comment:
So we have to make a copy of the message even for non-matching group ids, because we can't apply privacy deobfuscation to get the group id without making the copy?

I wonder whether it's worth, as a followup, seeing whether we can copy just the privacyheader parts.